content/chapters/compute/lab/README.md

```diff
@@ -0,0 +1,201 @@
+In our case, the operation simply adds the elements to an accumulator: `a + b`.
+Follow and run the code in `sum_array_threads_reduce.d`.
+
+The number of threads is used within a [`TaskPool`](https://dlang.org/phobos/std_parallelism.html#.TaskPool)).
+This sturcrure is a thread manager (not scheduler).
+It silently creates the number of threads we request and then `reduce` spreads its workload between these threads.
+
+Now run the `sum_array_threads_reduce` binary using 1, 2, 4, and 8 threads as before.
+You'll see lower running times than `sum_array_threads` due to the highly-optimised code of the `reduce` function.
+For this reason and because library functions are usually much better tested than your own code, it is always preferred to use a library function for a given task.
+
+### Threads vs Processes
+
+So why use the implementation that spawns more processes if it's slower than the one using threads?
+
+TODO: security, isolation, separate tasks
+
+## Arena
+
+### Threads and Processes: `clone`
+
+Let's go back to our initial demos that used threads and processes.
+We'll see that in order to create both threads and processes, the underlying Linux syscall is `clone`.
+For this, we'll run both `sum_array_threads` and `sum_array_processes` under `strace`.
+As we've already established, we're only interested in the `clone` syscall:
+```
+student@os:~/.../lab/support/sum-array/d/$ strace -e clone ./sum_array_threads 2
+clone(child_stack=0x7f60b56482b0, flags=CLONE_VM|CLONE_FS|CLONE_FILES|CLONE_SIGHAND|CLONE_THREAD|CLONE_SYSVSEM|CLONE_SETTLS|CLONE_PARENT_SETTID|CLONE_CHILD_CLEARTID, parent_tid=[1819693], tls=0x7f60b5649640, child_tidptr=0x7f60b5649910) = 1819693
+clone(child_stack=0x7f60b4e472b0, flags=CLONE_VM|CLONE_FS|CLONE_FILES|CLONE_SIGHAND|CLONE_THREAD|CLONE_SYSVSEM|CLONE_SETTLS|CLONE_PARENT_SETTID|CLONE_CHILD_CLEARTID, parent_tid=[1819694], tls=0x7f60b4e48640, child_tidptr=0x7f60b4e48910) = 1819694
+
+student@os:~/.../lab/support/sum-array/d/$ strace -e clone ./sum_array_processes 2
+clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x7f7a4e346650) = 1820599
+clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x7f7a4e346650) = 1820600
+```
+
+We ran each program with an argument of 2, so we have 2 calls to `clone`.
+Notice that in the case of threads, then `clone` syscall receives more arguments.
+If we check [`clone`'s man page](https://man.archlinux.org/man/clone3.2.en) `flags` it receives specify the following:
+- `CLONE_VM`: the child and the parent process share the same VAS
+- `CLONE_{FS,FILES,SIGHAND}`: the new thread shares the filesystem information, file and signal handlers with the one that created it
+The syscall also receives valid pointers to the new thread's stack and TLS, i.e. the only parts of the VAS that are distinct between threads (although they are technically accessible from all threads).
```

Comment:
This is a rather out-of-nowhere reference to TLS. I assume it will be detailed sometime in the lab text.

Reply:
Yes. I think this is alright since this section comes at the end of the "regular" lab.